core/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java

```
@@ -0,0 +1,193 @@
+            @Override
+            public void beforeBulk(long executionId, BulkRequest request) {
+                // no op
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+                responses.add(response);
+                latch.countDown();
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+                responses.add(failure);
+                latch.countDown();
+            }
+        }).setBulkActions(1)
+                .setConcurrentRequests(100)
+                .build();
+        indexDocs(bulkProcessor, numberOfAsyncOps);
+        latch.await(10, TimeUnit.SECONDS);
+        bulkProcessor.close();
+
+        assertThat(responses.size(), equalTo(numberOfAsyncOps));
+
+        // validate all responses
+        for (Object response : responses) {
+            if (response instanceof BulkResponse) {
+                BulkResponse bulkResponse = (BulkResponse) response;
+                for (BulkItemResponse bulkItemResponse : bulkResponse.getItems()) {
+                    if (bulkItemResponse.isFailed()) {
+                        BulkItemResponse.Failure failure = bulkItemResponse.getFailure();
+                        Throwable rootCause = ExceptionsHelper.unwrapCause(failure.getCause());
+                        // we're overwhelming the pool, that's the exception we want to get
+                        if (rootCause instanceof EsRejectedExecutionException == false) {
+                            fail("Got unexpected reason [" + failure.getMessage() + "]");
+                        }
+                    }
+                }
+            } else {
+                Throwable t = (Throwable) response;
+                // we're not expecting any other errors
+                fail("Got unexpected failure: " + t);
```

Throwable's toString eats stack traces. I'd just `throw new RuntimeException("unexpected failure", e);`